python/ql/src/experimental/Security/CWE-022bis/TarSlipImprov.ql

```diff
@@ -0,0 +1,108 @@
+/**
+ * @name Arbitrary file write during tarfile extraction
+ * @description Extracting files from a malicious tar archive without validating that the
+ *              destination file path is within the destination directory can cause files outside
+ *              the destination directory to be overwritten.
+ * @kind path-problem
+ * @id py/tarslip
+ * @problem.severity error
+ * @security-severity 7.5
+ * @precision high
+ * @tags security
+ *       external/cwe/cwe-022
+ */
+
+import python
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.TaintTracking
+import DataFlow::PathGraph
+import semmle.python.ApiGraphs
+import semmle.python.dataflow.new.internal.Attributes
+import semmle.python.dataflow.new.BarrierGuards
+import semmle.python.dataflow.new.RemoteFlowSources
+
+/**
+ * Handle those three cases of Tarfile opens:
+ *  - `tarfile.open()`
+ *  - `tarfile.TarFile()`
+ *  - `MKtarfile.Tarfile.open()`
+ */
+API::Node tarfileOpen() {
+  result in [
+      API::moduleImport("tarfile").getMember(["open", "TarFile"]),
+      API::moduleImport("tarfile").getMember("TarFile").getASubclass().getMember("open")
+    ]
+}
+
+/**
+ * Handle the previous three cases, plus the use of `closing` in the previous cases
+ */
+class AllTarfileOpens extends API::CallNode {
+  AllTarfileOpens() {
+    this = tarfileOpen().getACall()
+    or
+    exists(API::Node closing, Node arg |
+      closing = API::moduleImport("contextlib").getMember("closing") and
+      this = closing.getACall() and
+      arg = this.getArg(0) and
+      arg = tarfileOpen().getACall()
+    )
+  }
+}
```

Comment:
I would write this as a predicate like `fileOpen`.

I would also create a taint-step for the `closing`-call. For code like
```python
with closing(tarfile.open('example.tar', 'r')) as t:
    t.extractall('outdir')
```
we want a step from the argument of `closing()` to the call itself, whenever the argument is a value reachable from `tarfileOpen`. Then taint will automatically end up at `t`.

If you add such a taint-step, you only need `tarfileOpen()` as sources, and the paths will be nicer (but you still need `allTarfileOpen` to define sinks).

Reply:
So, if I understood well, you suggest that I simplify the Sources with an additional taint while keeping the AllTarfileOpens for the sinks, right?